Automattic/woocommerce-payments
# Pass previous Exception with Exception

#### Changes proposed in this Pull Request
While investigating successfully captured and failed orders root causes, I found we could use more information on error ['We're not able to process this request. Please refresh the page and try again.'](https://github.com/Automattic/woocommerce-payments/blob/5b9809465e0fb9ec3476d40f82b85daf82d512e9/includes/class-wc-payments-utils.php#L576-L579)

Changes made in this PR:
1. Passes the previous Exception with `Exception`.
2. Updates `Process_Payment_Exception` inside the `process_payment` to more unique Exceptions like `Fraud_Prevention_Enabled_Exception`, `Invalid_Phone_Number_Exception`, and `Rate_Limiter_Enabled_Exception` so when WC catches the Exception that throws from `process_payment`, we can pass the Exception class as additional data with the Exception to help WooPay better understanding where the errors come from.

#### Testing instructions
1. Test with [WooCommerce PR](https://github.com/woocommerce/woocommerce/pull/47489)
2. Return `true` in [check_nonce](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/StoreApi/Routes/V1/AbstractCartRoute.php#L285)
3. Run the PHP unit tests
4. Test through Postman for each Exception case inside `process_payment`:
    ( I tested them by commenting out the condition check )
   1. Add an item to your cart `{YOUR_DOMAIN}/wp-json/wc/store/v1/cart/add-item`
   
```
Body - Raw - JSON
{
  "id": 2040,
  "quantity": 1
}
```
    
     ii. Checkout with the cart `{YOUR_DOMAIN}/wp-json/wc/store/v1/checkout`
   
 
```
{
  "billing_address": {
        "first_name": "Test",
        "last_name": "Billing",
        "company": "",
        "address_1": "123 test",
        "address_2": "",
        "city": "San Diego",
        "state": "CA",
        "postcode": "92103",
        "country": "US",
        "email": “test@test.com”,
        "phone": "+14708274627"
    },
    "shipping_address": {
        "first_name": "Test",
        "last_name": “Shipping”,
        "company": "",
        "address_1": "123 test",
        "address_2": "",
        "city": "San Diego",
        "state": "CA",
        "postcode": "92103",
        "country": "US",
        "email": "test@test.com",
        "phone": "+14708274627"
    },
  "payment_method": "woocommerce_payments"
}
```
5. See previous exception class in response




*

-------------------


**Post merge**
